gctscript/vm/gctscript.go

```diff
@@ -0,0 +1,82 @@
+package vm
+
+import (
+	"fmt"
+
+	"github.com/gofrs/uuid"
+	"github.com/thrasher-corp/gocryptotrader/gctscript/wrappers/validator"
+	log "github.com/thrasher-corp/gocryptotrader/logger"
+)
+
+// New returns a new instance of VM
+func New() *VM {
+	if VMSCount.Len() >= int32(GCTScriptConfig.MaxVirtualMachines) {
```

Comment:
Something could theoretically hit this at the exact same time and allow the creation of more VMs than allowed.
eg
- VMS len at 3, max vms at 4
- New called simultaneously 5 times
- VMSCount  not yet >= 4
- New VM created and stored 5 times
- VMS count now added
- Now `VMSCount` can't be surpassed.

A lock or `VMSCount.add()` immediately after the if would solve this since there are no errs that could happen with 
```
vm := NewVM()
AllVMSync.Store(vm.ID, vm)
```

Reply:
I made this change moved the add further up also had to add logic to remove if NewVM returns nil

Reply:
I still had a feeling that the atomics wouldn't 100% protect an extra vm from spawning, but I wrote a test which has assured me that its not something to be concerned by:
```
func TestEverything(t *testing.T) {
	testRuns := 500
	newChecks := 10000
	for i := 0; i < testRuns; i++ {
		t.Run(strconv.FormatInt(int64(i), 10), func(t *testing.T) {
			for i := 0; i < newChecks; i++ {
				go New()
				go New()
				go New()
			}
			if VMSCount.Len() != int32(maxTestVirtualMachines) {
				t.Errorf("Success fail - Expected %v got %v", maxTestVirtualMachines, VMSCount.Len())
			}
			t.Log(VMSCount.Len())
		})
	}
}
```